hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/ReplicatedMapSplitBrainHandlerService.java

```diff
@@ -90,17 +106,136 @@ public Runnable prepareMergeRunnable() {
+                if (entries == null) {
+                    entries = new LinkedList<SplitBrainMergeEntryView<Object, Object>>();
+                    entriesPerPartition[partitionId] = entries;
+                }
+
+                SplitBrainMergeEntryView<Object, Object> entryView = createSplitBrainMergeEntryView(record);
+                entries.add(entryView);
+
+                long currentSize = ++counterPerMember[partitionId].value;
+                if (currentSize % batchSize == 0) {
+                    List<Integer> partitions = memberPartitionsMap.get(addresses[partitionId]);
+                    sendBatch(name, partitions, entriesPerPartition, mergePolicy);
+                }
+            }
+            // invoke operations for remaining entriesPerPartition
+            for (Map.Entry<Address, List<Integer>> entry : memberPartitionsMap.entrySet()) {
+                sendBatch(name, entry.getValue(), entriesPerPartition, mergePolicy);
+            }
+            return recordCount;
+        }
+
+        private void sendBatch(String name, List<Integer> memberPartitions,
+                               List<SplitBrainMergeEntryView<Object, Object>>[] entriesPerPartition,
+                               SplitBrainMergePolicy mergePolicy) {
+            int size = memberPartitions.size();
+            int[] partitions = new int[size];
+            int index = 0;
+            for (Integer partitionId : memberPartitions) {
+                if (entriesPerPartition[partitionId] != null) {
+                    partitions[index++] = partitionId;
+                }
+            }
+            if (index == 0) {
+                return;
+            }
+            // trim partition array to real size
+            if (index < size) {
+                partitions = Arrays.copyOf(partitions, index);
+                size = index;
+            }
+
+            //noinspection unchecked
+            List<SplitBrainMergeEntryView<Object, Object>>[] entries = new List[size];
+            index = 0;
+            int totalSize = 0;
+            for (int partitionId : partitions) {
+                int batchSize = entriesPerPartition[partitionId].size();
+                entries[index++] = entriesPerPartition[partitionId];
+                totalSize += batchSize;
+                entriesPerPartition[partitionId] = null;
+            }
+            if (totalSize == 0) {
+                return;
+            }
+
+            invokeMergeOperationFactory(name, mergePolicy, partitions, entries, totalSize);
+        }
+
+        private void invokeMergeOperationFactory(String name, SplitBrainMergePolicy mergePolicy, int[] partitions,
+                                                 List<SplitBrainMergeEntryView<Object, Object>>[] entries, int totalSize) {
+            try {
+                OperationFactory factory = new MergeOperationFactory(name, partitions, entries, mergePolicy);
+                nodeEngine.getOperationService()
```

Comment:
how does this part work if ownership of a partition is moved from one node to another in the mean time?

Reply:
That factory has a built-in support for failed partitions, which are then retried as single operation. This batching logic is stolen from the map.putAll() optimization I've done some versions ago. I'm on my mobile right now, but if you look into it you'll see a createPartionOperation() method or so, which is called for the failed ones by the operation service.